py-polars/polars/internals/series/utils.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import inspect
+import sys
+from functools import wraps
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+import polars.internals as pli
+from polars.datatypes import DataType, dtype_to_ffiname
+
+if TYPE_CHECKING:
+    from polars.polars import PySeries
+
+    if sys.version_info >= (3, 10):
+        from typing import ParamSpec
+    else:
+        from typing_extensions import ParamSpec
+
+    T = TypeVar("T")
+    P = ParamSpec("P")
+    SeriesMethod = Callable[..., pli.Series]
+
+
+class _EmptyBytecodeHelper:
+    def __init__(self) -> None:
+        # generate bytecode for empty functions with/without a docstring
+        def _empty_with_docstring() -> None:
+            """"""
+
+        def _empty_without_docstring() -> None:
+            pass
+
+        self.empty_bytecode = (
+            _empty_with_docstring.__code__.co_code,
+            _empty_without_docstring.__code__.co_code,
+        )
+
+    def __contains__(self, item: bytes) -> bool:
+        return item in self.empty_bytecode
+
+
+_EMPTY_BYTECODE = _EmptyBytecodeHelper()
+
+
+def _is_empty_method(func: SeriesMethod) -> bool:
+    """
+    Confirm that the given function has no implementation, eg:
+    * only has a docstring (body is empty)
+    * has no docstring and just contains 'pass' (or equivalent)
+    """
+    fc = func.__code__
+    return (fc.co_code in _EMPTY_BYTECODE) and (
+        len(fc.co_consts) == 2 and fc.co_consts[1] is None
+    )
+
+
+def _expr_lookup(namespace: str | None) -> set[tuple[str | None, str, tuple[str, ...]]]:
+    """Create lookup of potential Expr methods (in the given namespace)"""
+    # dummy Expr object that we can introspect
+    expr = pli.Expr()
+    expr._pyexpr = None
+
+    # optional indirection to "expr.str", "expr.dt", etc
+    if namespace is not None:
+        expr = getattr(expr, namespace)
+
+    lookup = set()
+    for name in dir(expr):
+        if not name.startswith("_"):
+            m = getattr(expr, name)
+            if callable(m):
+                # add function signature (argument names only) to the lookup
+                # as a _possible_ candidate for expression-dispatch
+                args = m.__code__.co_varnames[: m.__code__.co_argcount]
+                lookup.add((namespace, name, args))
+    return lookup
+
+
+def call_expr(func: SeriesMethod) -> SeriesMethod:
+    """Dispatch Series method to an expression implementation."""
+
+    @wraps(func)  # type: ignore[arg-type]
+    def wrapper(self: Any, *args: P.args, **kwargs: P.kwargs) -> pli.Series:
+        s = pli.wrap_s(self._s)
+        expr = pli.col(s.name)
+        namespace = getattr(self, "_accessor", None)
+        if namespace is not None:
+            expr = getattr(expr, namespace)
+        f = getattr(expr, func.__name__)
+        return s.to_frame().select(f(*args, **kwargs)).to_series()
```

Comment:
This could possibly be altered in order to preserve the class of `self` in the case of end-users sub-classing `Series`:

```suggestion
    @wraps(func)  # type: ignore[arg-type]
    def wrapper(self: Any, *args: P.args, **kwargs: P.kwargs) -> pli.Series:
        s = self._from_pyseries(self._s)
        expr = pli.col(s.name)
        namespace = getattr(self, "_accessor", None)
        if namespace is not None:
            expr = getattr(expr, namespace)
        f = getattr(expr, func.__name__)
        return self._from_pyseries(
            s.to_frame().select(f(*args, **kwargs)).to_series()._s
        )
```

In that case perhaps `wrapper`'s return type should be annotated with `typing.Self` (python 3.11+) or `typing_extensions.Self` (python <= 3.10). What do you think?

Reply:
Good catch! Now you've drawn my attention to it, I notice that `_from_pyseries` isn't actually accessible from the `*NameSpace` classes, resulting in the following exception:

```python
class ExtendedSeries(pl.Series):
    """Custom Series class"""

s = ExtendedSeries("s", ["x!", "y!", "z?"])
res = s.str.contains("?", literal=True)

# AttributeError: 'StringNameSpace' object has no attribute '_from_pyseries'
```

Further... it doesn't actually seem like `_from_pyseries` is used/working for a large number of existing (non-decorated) functions anyway? I may be misunderstanding its scope, but it seems the intent is to preserve the user-defined class in the general case.

```python
class XSeries(pl.Series):
    """Custom Series class"""

s = XSeries("s", [1, 2, 3])

# for example, all of the following return Series instead of XSeries
s.reinterpret(False)
s.shrink_to_fit()
s.hash()
s.abs()

# ...and so on

Reply:
Just made a small Series `repr` update that could help show this a bit more clearly, by including the custom series class name: https://github.com/pola-rs/polars/pull/4571

Reply:
> In that case perhaps wrapper's return type could (should?) be annotated with typing.Self (python 3.11+) or typing_extensions.Self (python <= 3.10). What do you think?

Unfortunately, mypy doesn't support the `Self` type yet: https://github.com/python/mypy/issues/11871

But I think we can rewrite the decorator to preserve the output type of the original method.

Reply:
> So I propose we save this improvement for a future PR. This PR is already big enough! Maybe you could make a separate issue for this?

My thoughts exactly; it doesn't seem to work at the moment anyway, so no advantage holding off on this. I don't mind volunteering for the follow-up (I've already got a reasonable idea how it can be done with minimal changes) ;)